reps/2023-07-20-torch-trainer-apis.md

```diff
@@ -0,0 +1,371 @@
+# [Train] Unify Torch based Trainers on the `TorchTrainer` API
+
+## Summary
+
+Remove the `LightningTrainer`, `TransformersTrainer`, and `AccelerateTrainer` APIs and unify them on the `TorchTrainer` API.
+
+### General Motivation
+
+Ray Train currently exposes a different `Trainer` API for each individual training framework. The PyTorch based Trainers are as follows:
+
+| Framework | Ray Train Trainer |
+| ----------- | ----------- |
+| PyTorch | `TorchTrainer` |
+| Lightning | `LightningTrainer` |
+| HuggingFace Transformers | `TransformersTrainer` |
+| HuggingFace Accelerate | `AccelerateTrainer` |
+
+While this pattern makes it very obvious to the user that Ray Train provides an integration with a given framework, we have observed a few drawbacks:
+1. There is overhead for the user to learn and use these APIs.
+    1. They are all different from each other.
+    2. They are different from their corresponding framework.
+    3. The user cannot see the internal training loop, which makes it hard to implement and debug their code.
+2. The `LightningTrainer` and `TransformersTrainer` APIs are opionated and may not allow the user to fully express their desired training logic (e.g. validation and testing).
+
+This proposal explores the idea of centralizing on a single `TorchTrainer` as the single way running training code for PyTorch-based frameworks in a distributed fashion.
```

Comment:
I also want to call out that we do this for TensorflowTrainer (Native TF vs. Keras) already (though the surface area is smaller with only 2 APIs).